viewing_party/party.py

```diff
@@ -1,23 +1,121 @@
+# ---------------------------------------------------- WAVE 1 ----------------------------------------------------
+#-----------------------------------------------------------------------------------------------------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if not title or not genre or not rating:
+        return None
+    
+    new_movie_info = {"title": title, "genre": genre, "rating": rating}
+    return new_movie_info
 
-# -----------------------------------------
-# ------------- WAVE 2 --------------------
-# -----------------------------------------
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
 
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+    return user_data
 
-        
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+# ---------------------------------------------------- WAVE 2 ----------------------------------------------------
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0
+    
+    total_movie_ratings = 0
 
+    for movie_rating in user_data["watched"]:
+        total_movie_ratings += movie_rating["rating"]
+
+    average_movie_rating = total_movie_ratings / len(user_data["watched"])
+
+    return average_movie_rating
+
+def get_most_watched_genre(user_data):
+    frequency_of_genre_index = {}
+    max_frequency = 0
+    most_watched_genre = None
+
+    for single_movie in user_data["watched"]:
+        genre_type = single_movie.get("genre")
+        if genre_type:
+            frequency_of_genre_index[genre_type] = frequency_of_genre_index.get(genre_type,0) + 1
+            if frequency_of_genre_index[genre_type] > max_frequency:
+              max_frequency = frequency_of_genre_index[genre_type]
+              most_watched_genre = genre_type
+
+    return most_watched_genre
+
+# ---------------------------------------------------- WAVE 3 ----------------------------------------------------
+
+def get_unique_watched(user_data):
+    if not user_data["watched"]:
+        return []
+  
+    my_friends_unique = [movie["title"] for my_friends in user_data["friends"] 
+                         for movie in my_friends["watched"]]
+    
+    my_unique_movies = [movie for movie in user_data["watched"] 
+                        if movie["title"] not in my_friends_unique]   

```

Comment:
question: will ask about it in office hours

Reply:
A set can take a constant amount of time to find the elements. A list would have to look at n-1.

Reply:
in my_unique_movies....

if movie["title"] not in my_friends_unique]   

searching through the my_friends_unique variable would take a constant if it was a set rather than if it was a list.